### PR TITLE
support clean up the historical control_directory 

### LINKED ISSUE #125: Is it possible to clean up the historical control_directory before creating a SessionBuilder?
Currently testing locally, the historical [control_directory](https://docs.rs/openssh/latest/openssh/struct.SessionBuilder.html#method.control_directory) directory will always exist during each debugging session.

E.g:
```shell
➜ ll -als
total 0
0 drwxr-xr-x  6 root  staff   192B  7 28 10:03 .
0 drwxr-xr-x  3 root  staff    96B  7 27 18:25 ..
0 drwxr-xr-x  4 root  staff   128B  7 28 09:55 .ssh-connectionpCn9J7
0 drwxr-xr-x  2 root  staff    64B  7 28 10:01 .ssh-connectionpHq9a2
0 drwxr-xr-x  2 root  staff    64B  7 28 10:03 .ssh-connectionpJp983
0 drwxr-xr-x  2 root  staff    64B  7 28 10:05 .ssh-connectionpLwp24
```

### PR DESCRIPTION
**Pull request by baoyachi**
fix https://github.com/openssh-rust/openssh/issues/125

### CODE REVIEW

**Comment by jonhoo** [comment]
This change is [Reviewable](https://reviewable.io/reviews/openssh-rust/openssh/126)

File: src/builder.rs [hunk 1]
```diff
@@ -34,6 +34,31 @@ fn get_default_control_dir<'a>() -> Result<&'a Path, Error> {
         })
 }
 
+fn clean_history_control_dir(dir: &TempDir, prefix: &str) -> io::Result<()> {
+    // Check if the parent directory of the given TempDir exists
+    if let Some(parent) = dir.path().parent() {
+        // Read the entries in the parent directory
+        fs::read_dir(parent)?
+            // Filter out and keep only the valid entries
+            .filter_map(|entry| entry.ok())
```
Comment (NobodyXu) [inline src/builder.rs:43]:
```suggestion
            .filter_map(Result::ok)
```

File: src/builder.rs [hunk 2]
```diff
@@ -133,6 +160,24 @@ impl SessionBuilder {
...
+    ///
+    /// Generally, the configuration of 'control_directory' is usually a Temporary folder.
+    /// Please refer to issue:https://github.com/openssh-rust/openssh/pull/28
+    //
+    // If not configured, openssh will automatically create a temporary .ssh_connection dir in the
+    // current running environment.
+    //
+    // This method provides control for cleaning up history 'control_directory' function,
+    // default openssh will not clean up. If necessary, you can manually call this method
+    ///
```
Comment (NobodyXu) [inline src/builder.rs:173]:
```suggestion
    /// Clean up the temporary directories with the `.ssh-connection` prefix
    /// in directory specified by [`SessionBuilder::control_directory`], created by
    /// previous `openssh::Session` that is not cleaned up for some reasons
    /// (e.g. process getting killed, abort on panic, etc)
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/builder.rs [hunk 3]
```diff
@@ -133,6 +160,17 @@ impl SessionBuilder {
         self
     }
 
+    /// Clean up the temporary directories with the `.ssh-connection` prefix
+    /// in directory specified by [`SessionBuilder::control_directory`], created by
+    /// previous `openssh::Session` that is not cleaned up for some reasons
+    /// (e.g. process getting killed, abort on panic, etc)
```
Comment (NobodyXu) [inline src/builder.rs:166]:
```suggestion
    /// (e.g. process getting killed, abort on panic, etc)
    ///
    /// Use this with caution, do not enable this if you don't understand
    /// what it does,
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/builder.rs [hunk 4]
```diff
@@ -295,11 +336,16 @@ impl SessionBuilder {
...
+        let prefix = ".ssh-connection";
         let dir = Builder::new()
-            .prefix(".ssh-connection")
+            .prefix(prefix)
             .tempdir_in(socketdir)
             .map_err(Error::Master)?;
 
+        if self.clean_history_control_dir {
+            let _ = clean_history_control_dir(&dir, prefix);
+        }
```
Comment (NobodyXu) [inline src/builder.rs:347]: You can simplify this by calling `clean_history_control_dir` before creating the new `TempDir`, then you won't have to compare the `entry.path() != dir.path()` and you would also don't need to call `dir.path().parent()` with a `if-let`

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/builder.rs (hunk 4 above)
Comment (baoyachi) [reply src/builder.rs:347]: That's a good suggestion.

**Comment by NobodyXu** [comment]
P.S. I would recommend rebasing against main instead of merging main into this PR, which will create an additional commit.

**Review by NobodyXu — APPROVED** [review]

**Comment by jonhoo** [comment]
I assume these mostly stuck around for when a connection failed, which has since been fixed (I think). I'm not in love with doing a recursive directory remove here invisibly on the user's behalf — feels like a recipe for accidentally deleting something unintentional down the line.

**Comment by NobodyXu** [comment]
> I assume these mostly stuck around for when a connection failed, which has since been fixed (I think). I'm not in love with doing a recursive directory remove here invisibly on the user's behalf — feels like a recipe for accidentally deleting something unintentional down the line.

This behavior is disabled by default, so I think it might be fine.

**Pull request merged**